File: kodibot/telegram/rate.py
"""Rate-limited Telegram API wrappers and message-tracking helpers.

Provides ``telegram_request`` / ``telegram_request_delete`` for serialised,
rate-limited Telegram API calls, plus convenience helpers like
``send_and_track`` and ``delete_message_if_present``.
"""

import asyncio
import logging
import time

from telegram.error import RetryAfter, TimedOut

from kodibot.telegram import state as S

log = logging.getLogger(__name__)


async def telegram_request(call, *args, **kwargs):
    """Serialise a Telegram API call with rate-limiting and retry."""
    for _ in range(S.TG_MAX_RETRIES):
        async with S.TG_RATE_LOCK:
            now = time.time()
            extra = S.TG_DYNAMIC_DELAY if now < S.TG_DYNAMIC_UNTIL else 0.0
            wait = S.TG_MIN_INTERVAL + extra - (now - S.TG_LAST_TS)
            if wait > 0:
                await asyncio.sleep(wait)
            try:
                res = await call(*args, **kwargs)
                S.TG_LAST_TS = time.time()
                return res
            except RetryAfter as e:
                S.TG_LAST_TS = time.time()
                S.TG_DYNAMIC_DELAY = max(S.TG_DYNAMIC_DELAY, min(e.retry_after, 2.0))
                S.TG_DYNAMIC_UNTIL = time.time() + 60.0
                await asyncio.sleep(e.retry_after)
            except TimedOut:
                S.TG_LAST_TS = time.time()
                await asyncio.sleep(1.5)
            except Exception:
                S.TG_LAST_TS = time.time()
                raise
    # Final attempt without catching
    async with S.TG_RATE_LOCK:
        now = time.time()
        extra = S.TG_DYNAMIC_DELAY if now < S.TG_DYNAMIC_UNTIL else 0.0
        wait = S.TG_MIN_INTERVAL + extra - (now - S.TG_LAST_TS)
        if wait > 0:
            await asyncio.sleep(wait)
        res = await call(*args, **kwargs)
        S.TG_LAST_TS = time.time()
        return res


async def telegram_request_delete(call, *args, **kwargs):
    """Serialise a Telegram delete API call with separate rate-limiting."""
    for _ in range(S.TG_MAX_RETRIES):
        async with S.TG_DELETE_RATE_LOCK:
            now = time.time()
            extra = S.TG_DYNAMIC_DELAY if now < S.TG_DYNAMIC_UNTIL else 0.0
            wait = S.TG_DELETE_MIN_INTERVAL + extra - (now - S.TG_DELETE_LAST_TS)
            if wait > 0:
                await asyncio.sleep(wait)
            try:
                res = await call(*args, **kwargs)
                S.TG_DELETE_LAST_TS = time.time()
                return res
            except RetryAfter as e:
                S.TG_DELETE_LAST_TS = time.time()
                S.TG_DYNAMIC_DELAY = max(S.TG_DYNAMIC_DELAY, min(e.retry_after, 2.0))
                S.TG_DYNAMIC_UNTIL = time.time() + 60.0
                await asyncio.sleep(e.retry_after)
            except TimedOut:
                S.TG_DELETE_LAST_TS = time.time()
                await asyncio.sleep(1.5)
            except Exception:
                S.TG_DELETE_LAST_TS = time.time()
                raise
    async with S.TG_DELETE_RATE_LOCK:
        now = time.time()
        extra = S.TG_DYNAMIC_DELAY if now < S.TG_DYNAMIC_UNTIL else 0.0
        wait = S.TG_DELETE_MIN_INTERVAL + extra - (now - S.TG_DELETE_LAST_TS)
        if wait > 0:
            await asyncio.sleep(wait)
        res = await call(*args, **kwargs)
        S.TG_DELETE_LAST_TS = time.time()
        return res


async def send_and_track(ctx, chat_id, text, **kwargs):
    """Send a message and track its message_id in shared state."""
    if "disable_web_page_preview" not in kwargs:
        kwargs["disable_web_page_preview"] = True
    msg = await telegram_request(ctx.bot.send_message, chat_id=chat_id, text=text, **kwargs)
    if chat_id not in S.FIRST_BOT_ID:
        S.FIRST_BOT_ID[chat_id] = msg.message_id
    S.PREV_BOT_ID[chat_id] = S.LAST_BOT_ID.get(chat_id)
    S.LAST_BOT_ID[chat_id] = msg.message_id
    log.debug("Bot msg chat_id=%s message_id=%s", chat_id, msg.message_id)
    return msg


async def delete_message_if_present(ctx, chat_id, message_id):
    """Silently delete a message (or list of message ids)."""
    if not message_id:
        return
    if isinstance(message_id, (list, tuple, set)):
        for mid in message_id:
            await delete_message_if_present(ctx, chat_id, mid)
        return
    try:
        await telegram_request_delete(ctx.bot.delete_message, chat_id=chat_id, message_id=message_id)
    except Exception:
        pass

File: internal/telegram/rate.go
package telegram

import (
	"log"
	"sync"
	"time"

	"github.com/PaulSonOfLars/gotgbot/v2"
)

// ── Rate limiting constants ─────────────────────────────────────────

const (
	tgMinInterval       = 0.06  // 60ms between API calls
	tgDeleteMinInterval = 0.06
	tgMaxRetries        = 3
)

var (
	tgLastTS         float64
	tgDeleteLastTS   float64
	tgDynamicDelay   float64
	tgDynamicUntil   float64
	tgRateLock       sync.Mutex
	tgDeleteRateLock sync.Mutex
)

// TelegramRequest wraps a Telegram API call with rate limiting and retry.
// The caller passes a function that performs the actual API call.
func TelegramRequest(fn func() error) error {
	for attempt := 0; attempt < tgMaxRetries; attempt++ {
		tgRateLock.Lock()
		now := float64(time.Now().UnixMilli()) / 1000.0
		extra := 0.0
		if now < tgDynamicUntil {
			extra = tgDynamicDelay
		}
		wait := tgMinInterval + extra - (now - tgLastTS)
		if wait > 0 {
			time.Sleep(time.Duration(wait * float64(time.Second)))
		}
		err := fn()
		tgLastTS = float64(time.Now().UnixMilli()) / 1000.0
		tgRateLock.Unlock()

		if err == nil {
			return nil
		}
		// Check for rate limiting / timeout errors
		log.Printf("WARN Telegram request error (attempt %d/%d): %v", attempt+1, tgMaxRetries, err)
		time.Sleep(1500 * time.Millisecond)
	}
	// Final attempt
	tgRateLock.Lock()
	now := float64(time.Now().UnixMilli()) / 1000.0
	extra := 0.0
	if now < tgDynamicUntil {
		extra = tgDynamicDelay
	}
	wait := tgMinInterval + extra - (now - tgLastTS)
	if wait > 0 {
		time.Sleep(time.Duration(wait * float64(time.Second)))
	}
	err := fn()
	tgLastTS = float64(time.Now().UnixMilli()) / 1000.0
	tgRateLock.Unlock()
	return err
}

// TelegramRequestDelete wraps a Telegram delete API call with separate rate limiting.
func TelegramRequestDelete(fn func() error) error {
	for attempt := 0; attempt < tgMaxRetries; attempt++ {
		tgDeleteRateLock.Lock()
		now := float64(time.Now().UnixMilli()) / 1000.0
		extra := 0.0
		if now < tgDynamicUntil {
			extra = tgDynamicDelay
		}
		wait := tgDeleteMinInterval + extra - (now - tgDeleteLastTS)
		if wait > 0 {
			time.Sleep(time.Duration(wait * float64(time.Second)))
		}
		err := fn()
		tgDeleteLastTS = float64(time.Now().UnixMilli()) / 1000.0
		tgDeleteRateLock.Unlock()

		if err == nil {
			return nil
		}
		log.Printf("WARN Telegram delete error (attempt %d/%d): %v", attempt+1, tgMaxRetries, err)
		time.Sleep(1500 * time.Millisecond)
	}
	// Final attempt
	tgDeleteRateLock.Lock()
	now := float64(time.Now().UnixMilli()) / 1000.0
	extra := 0.0
	if now < tgDynamicUntil {
		extra = tgDynamicDelay
	}
	wait := tgDeleteMinInterval + extra - (now - tgDeleteLastTS)
	if wait > 0 {
		time.Sleep(time.Duration(wait * float64(time.Second)))
	}
	err := fn()
	tgDeleteLastTS = float64(time.Now().UnixMilli()) / 1000.0
	tgDeleteRateLock.Unlock()
	return err
}

// SendAndTrack sends a message and tracks its ID.
func SendAndTrack(bot *gotgbot.Bot, chatID int64, text string, opts *gotgbot.SendMessageOpts) (*gotgbot.Message, error) {
	if opts == nil {
		opts = &gotgbot.SendMessageOpts{}
	}
	if opts.LinkPreviewOptions == nil {
		opts.LinkPreviewOptions = &gotgbot.LinkPreviewOptions{IsDisabled: true}
	}

	var msg *gotgbot.Message
	err := TelegramRequest(func() error {
		var e error
		msg, e = bot.SendMessage(chatID, text, opts)
		return e
	})
	if err != nil {
		return nil, err
	}
	if msg != nil {
		StateMu.Lock()
		PrevBotID = LastBotID
		LastBotID = msg.MessageId
		ActiveChatID = chatID
		StateMu.Unlock()
		log.Printf("DEBUG Bot msg chat_id=%d message_id=%d", chatID, msg.MessageId)
	}
	return msg, nil
}

// DeleteMessageIfPresent silently deletes a message.
func DeleteMessageIfPresent(bot *gotgbot.Bot, chatID int64, messageID int64) {
	if messageID == 0 {
		return
	}
	_ = TelegramRequestDelete(func() error {
		_, err := bot.DeleteMessage(chatID, messageID, nil)
		return err
	})
}

// DeleteMessagesIfPresent silently deletes multiple messages.
func DeleteMessagesIfPresent(bot *gotgbot.Bot, chatID int64, messageIDs []int64) {
	for _, mid := range messageIDs {
		DeleteMessageIfPresent(bot, chatID, mid)
	}
}
